### ISSUE TITLE
try two or three different interim priors

### ISSUE THREAD

**Issue by davidwhogg** [aimalz/prob-z#37]
Show that we get the right answer for the maximum marginalized likelihood, no matter what interim prior we use.  That is, choose three (say) interim priors:  One flat, one that slopes so that it prefers low redshifts, one curved (so it prefers, say, low and high over middle).  Show that the p(z) data (zPDFs) change sensibly as you change the interim prior, but that the final maximum marginalized likelihood doesn't change.

This is critical, because we are going to have _real_ situations with strange interim priors (think crazy training sets...).

**Comment by aimalz** [comment]
As measured by the log likelihood and KLD in comparison with the alternative methods, the maximum marginalized likelihood estimator is a "best fit" for all three interim priors.  EDIT: The maximum marginalized likelihood estimator does not change for each interim prior.  I was misreading my own output!
Flat:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11667261/a1c5ff32-9dbe-11e5-802a-94ec5c9b0707.png)
Low-z favoring:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11667279/b89701fc-9dbe-11e5-87fc-8f2799f365bc.png)
Mid-z disfavoring:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11667300/d476f5a8-9dbe-11e5-9cb3-2547d809ca0a.png)

**Comment by aimalz** [comment]
And the zPDFs do change as expected under these interim priors.  In each plot the true redshift is indicated by a dashed line and the "observed" redshift is indicated by a dotted line in the same color.
Flat:
![samplepzs](https://cloud.githubusercontent.com/assets/8606810/11668217/f6eb6fd8-9dc3-11e5-866d-824e31e25e9f.png)
Low-z favoring:
![samplepzs](https://cloud.githubusercontent.com/assets/8606810/11668224/007b9b86-9dc4-11e5-9f84-807062bfd2fa.png)
Mid-z disfavoring:
![samplepzs](https://cloud.githubusercontent.com/assets/8606810/11668234/0b58ff4e-9dc4-11e5-89fe-295c9131a098.png)

**Comment by davidwhogg** [comment]
Okay this is a start.  Now can you run the same tests but with 10x as many galaxies in the mix?  Then the dotted line should match the true line, right?

Also, plot the "truth" even thicker but with alpha=0.25, so we can see the other lines in that context.

Also, by MLE, I presume you mean maximum _marginalized_ likelihood -- ie, marginalizing out the true redshifts?  Or am I reading that wrong?

I don't like these acronyms -- let's not use them; let's use words, or else give a caption for every figure explaining these acronyms.

**Comment by aimalz** [comment]
The MLE marginalizes out the individual redshift posteriors here.  By eye, the dotted line (MLE) approaches the thicker line (true N(z)) with a larger survey size.  However, the KLD indicates that it gets worse.
Flat interim prior:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11673893/37231602-9dea-11e5-8f1b-19e438a8979f.png)
Low-z favoring interim prior:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11673911/5427588a-9dea-11e5-9a6c-2abef29f71bc.png)
Mid-z disfavoring interim prior:
![truenz](https://cloud.githubusercontent.com/assets/8606810/11673927/682ef5fe-9dea-11e5-87b7-05a3ffa5e3e4.png)

**Comment by davidwhogg** [comment]
It's not obviously getting better.  How long did these take, and can you afford to go up by another factor of 10?

**Comment by aimalz** [comment]
The optimization step is linear in number of galaxies, so it should take about 10 minutes for 10 times more galaxies.  (EDIT: But generating the data also takes longer by a linear amount, so it's actually a lot longer. . .)  I'll start it now.

**Comment by davidwhogg** [comment]
It takes a long time to generate the fake data?

**Comment by aimalz** [comment]
It takes about 5 s for 1000 galaxies in 5 bins and increases linearly with number of galaxies.  Calculating the MLE takes about 2 seconds for the same number of galaxies and bins (it was closer to 5 minutes with the optimizer I was using at the time the previous comment was posted) and increases linearly with the number of galaxies as well and as the square of the number of bins.

**Comment by davidwhogg** [comment]
Generating fake data can't possibly take that long.  You are drawing from a 1-d mixture of 5 Gaussians and then drawing from a second gaussian (to get the observed) and then evaluating some erf functions.  That can't take more than a fraction of a millisecond, right?

And I am asking about the data generation -- you said it takes a big fraction of the time.  If that is true, fix it.  Also, most experiments you do _don't_ requre that you re-generate the data.

**Comment by davidwhogg** [comment]
Also data generation can't depend on the number of bins -- or not strongly.  How are you doing the integrals over the bins?

**Comment by aimalz** [comment]
The slow steps in generating data involve getting PDF/CDF values from scipy probability distributions, because it has to be done over each bin separately.  Putting together that sum of Gaussians is taking longer than it should due to me being insufficiently pythonic, but the much larger time cost in making binned individual redshift posteriors probably can't be significantly reduced and that's what depends linearly on the number of bins.

Data generation isn't slow compared to MCMC, but it doesn't take a negligible amount of time either.  And, indeed, data does not need to be generated more than once for a given test, assuming the dimensionality and survey size are unchanged.  But data gets generated many times when I'm testing the data generation code!

EDIT: I found a way to make the construction of binned individual redshift posteriors take half as much time.  I also forgot to mention that another slow step is calculating the MMLE, which is done with a scipy.optimize routine.  That's now the slowest step.

**Comment by davidwhogg** [comment]
you should be able to do something like:

```
components = np.random.randint(0,5,size=n_gal)
chis = np.random.normal(size=n_gal)
truezs = means[components] + np.sqrt(vars[components]) * chis
```

No loops!

As for the CDF, write it in terms of `np.erf()` and see if that is far faster.

And never have any loop, ever.

**Comment by davidwhogg** [comment]
ps.  Even the CDF calculations can be done in single shots without loops over either bins or objects, I suspect.

**Comment by aimalz** [comment]
I wrote my own CDF/PDF functions that eliminate loops, and they're eight times faster than the scipy versions.  Thanks!  Now the only slow step is optimizing for the MLE.  It goes as number of bins squared and linearly in number of galaxies, so it can potentially be quite slow indeed.

**Comment by davidwhogg** [comment]
We are off topic.  Can you tell me if you think we have this issue satisfied, and repost the relevant figure that convinces me of such?

**Comment by aimalz** [comment]
I think this is resolved.  Here are the results of MCMC on a small dataset for flat, low-z favoring, and mid-z disfavoring interim priors.  The competing methods are strongly affected by the nontrivial interim priors, but the sampler is not.
![samps](https://cloud.githubusercontent.com/assets/8606810/11793188/2a29eca4-a278-11e5-9122-99bdc4b9761c.png)
![samps](https://cloud.githubusercontent.com/assets/8606810/11793162/0d62af16-a278-11e5-9aad-f1897d3189f3.png)
![samps](https://cloud.githubusercontent.com/assets/8606810/11793174/170fdcbe-a278-11e5-9af1-984664a10079.png)

**Comment by davidwhogg** [comment]
I don't understand -- these don't have the truth plotted correctly.  And I want "sampled" dropped.

**Comment by aimalz** [comment]
The truth in these cases is a histogram of the true redshifts from the simulation suite, rather than the function from which they were drawn.  I chose that format because it's a more generic data product that a user could plausibly provide.  It would be harder for me to rewrite the sampling suite to take in an arbitrary continuous function as truth from a user, although I could do it in my particular case for the purposes of writing the paper if you'd prefer.  

Sampled here refers to the accepted samples of parameter values produced by the sampler, which are plotted in color with a thicker linewidth for the best fits from each sequence of 1000 iterations.  (You asked to see a few random ones plotted, but I had a really hard time evaluating how good the samples were from just a few lines, so this was my compromise.)

**Comment by davidwhogg** [comment]
I have no idea what you are saying.  Even if the redshifts come from a simulation suite, they are drawn from a model that is not a five-bin step function!

I feel like I am saying the same things over and over:

Plot truth:  Thin, smooth line.  I am not sure why you are talking about a "user":  _You_ are the only user right now, or am I wrong?

Drop "sampled".  Never want to see it, except when we are debugging.

Most of the lines on the plot are completely unlabeled.  What are they?

Plot Maximum marginalized LF. Label it correctly.

Plot the histogram of interim MAP redshifts.  Label it correctly.

(Later we will plot the sum of interim posteriors.  But not right now.)

Plot the interim prior as a SMOOTH CURVE.  And make it a smooth curve.

I don't want to type these requests again, ever!

**Comment by aimalz** [comment]
Sorry for being dense.  I'll get right on it after the exam.
